File: WithApp/Views/Components/SupportMessageBubble.swift
//
//  SupportMessageBubble.swift
//  WithApp
//
//  Component for displaying supportive messages based on glucose levels
//

import SwiftUI

struct SupportMessageBubble: View {
    let message: SupportMessage
    @State private var isVisible = false
    
    var body: some View {
        VStack(spacing: 8) {
            Text(message.text)
                .font(.system(size: 15))
                .foregroundColor(.primary)
                .multilineTextAlignment(.center)
                .lineSpacing(4)
            
            Text("双击或上滑继续")
                .font(.system(size: 11))
                .foregroundColor(Color("AccentTeal").opacity(0.7))
        }
        .padding(.horizontal, 24)
        .padding(.vertical, 16)
        .background(
            MessageBubbleShape()
                .fill(Color.white)
                .shadow(color: .black.opacity(0.08), radius: 15, y: 5)
        )
        .opacity(isVisible ? 1 : 0)
        .offset(y: isVisible ? 0 : 10)
        .onAppear {
            withAnimation(.easeOut(duration: 0.5)) {
                isVisible = true
            }
        }
    }
}

// MARK: - Message Bubble Shape
struct MessageBubbleShape: Shape {
    func path(in rect: CGRect) -> Path {
        var path = Path()
        let cornerRadius: CGFloat = 20
        let arrowSize: CGFloat = 12
        let arrowX = rect.midX
        
        // Start from top left
        path.move(to: CGPoint(x: rect.minX + cornerRadius, y: rect.minY))
        
        // Top edge
        path.addLine(to: CGPoint(x: rect.maxX - cornerRadius, y: rect.minY))
        
        // Top right corner
        path.addQuadCurve(
            to: CGPoint(x: rect.maxX, y: rect.minY + cornerRadius),
            control: CGPoint(x: rect.maxX, y: rect.minY)
        )
        
        // Right edge
        path.addLine(to: CGPoint(x: rect.maxX, y: rect.maxY - cornerRadius))
        
        // Bottom right corner
        path.addQuadCurve(
            to: CGPoint(x: rect.maxX - cornerRadius, y: rect.maxY),
            control: CGPoint(x: rect.maxX, y: rect.maxY)
        )
        
        // Bottom edge with arrow
        path.addLine(to: CGPoint(x: arrowX + arrowSize, y: rect.maxY))
        path.addLine(to: CGPoint(x: arrowX, y: rect.maxY + arrowSize))
        path.addLine(to: CGPoint(x: arrowX - arrowSize, y: rect.maxY))
        path.addLine(to: CGPoint(x: rect.minX + cornerRadius, y: rect.maxY))
        
        // Bottom left corner
        path.addQuadCurve(
            to: CGPoint(x: rect.minX, y: rect.maxY - cornerRadius),
            control: CGPoint(x: rect.minX, y: rect.maxY)
        )
        
        // Left edge
        path.addLine(to: CGPoint(x: rect.minX, y: rect.minY + cornerRadius))
        
        // Top left corner
        path.addQuadCurve(
            to: CGPoint(x: rect.minX + cornerRadius, y: rect.minY),
            control: CGPoint(x: rect.minX, y: rect.minY)
        )
        
        return path
    }
}

// MARK: - Support Message Model
struct SupportMessage {
    let text: String
    let level: SupportLevel
    
    enum SupportLevel {
        case stable      // Gentle encouragement
        case borderline  // Soft reassurance
        case concerning  // Supportive prompt
    }
    
    static let defaultMessage = SupportMessage(
        text: "你的血糖在理想范围内。你现在感觉怎么样？",
        level: .stable
    )
    
    static func getMessage(for glucoseLevel: GlucoseLevel, context: ContextMode) -> SupportMessage {
        switch glucoseLevel {
        case .low:
            return SupportMessage(
                text: concerningMessages.randomElement() ?? defaultMessage.text,
                level: .concerning
            )
        case .borderlineLow, .borderlineHigh:
            return SupportMessage(
                text: borderlineMessages.randomElement() ?? defaultMessage.text,
                level: .borderline
            )
        case .normal:
            return contextAwareMessage(for: context)
        case .high:
            return SupportMessage(
                text: concerningMessages.randomElement() ?? defaultMessage.text,
                level: .concerning
            )
        case .unknown:
            return defaultMessage
        }
    }
    
    private static func contextAwareMessage(for context: ContextMode) -> SupportMessage {
        switch context {
        case .restaurant:
            return SupportMessage(
                text: "你现在像是在就餐场景。吃东西前，想先聊聊此刻的感受吗？",
                level: .stable
            )
        case .work:
            return SupportMessage(
                text: "工作间隙也值得给自己一点时间。你的状态很重要。",
                level: .stable
            )
        case .social:
            return SupportMessage(
                text: "社交时刻会有开心，也可能有压力。我在这里陪你。",
                level: .stable
            )
        default:
            return SupportMessage(
                text: stableMessages.randomElement() ?? defaultMessage.text,
                level: .stable
            )
        }
    }
    
    // MARK: - Message Collections
    private static let stableMessages = [
        "你的血糖在理想范围内。你现在感觉怎么样？",
        "今天整体比较平稳。想回顾一下今天的状态吗？",
        "身体状态不错，也想关心一下你的心情。",
        "给自己一个温柔的停顿。你今天的精力怎么样？",
        "数字很平稳。现在有没有什么在你心里？"
    ]
    
    private static let borderlineMessages = [
        "我注意到一些波动，不用紧张，我只是来关心你。",
        "身体正在调整。你此刻感觉如何？",
        "有波动是很常见的。想和我聊聊吗？",
        "温柔提醒一下：你已经很努力了。我可以怎么支持你？",
        "不管起伏如何，我都会陪着你。"
    ]
    
    private static let concerningMessages = [
        "我注意到一个需要留意的信号。你现在感觉怎么样？",
        "我们先一起慢下来。我会在这里支持你。",
        "你的整体状态最重要。愿意说说你现在的情况吗？",
        "我在这儿陪你。我们一起看看你此刻的感受。",
        "你不是一个人在面对这些。我可以怎么帮你？"
    ]
}

// MARK: - Preview
#Preview {
    VStack(spacing: 30) {
        SupportMessageBubble(message: .defaultMessage)
        
        SupportMessageBubble(message: SupportMessage(
            text: "我注意到一些波动，不用紧张，我只是来关心你。",
            level: .borderline
        ))
    }
    .padding()
    .background(Color(red: 0.95, green: 0.98, blue: 0.97))
}

File: app/src/main/java/com/withapp/with/ui/components/SupportMessageBubble.kt
//package com.withapp.with.ui.components
//
//import androidx.compose.animation.*
//import androidx.compose.foundation.background
//import androidx.compose.foundation.layout.*
//import androidx.compose.foundation.shape.RoundedCornerShape
//import androidx.compose.material3.Text
//import androidx.compose.runtime.*
//import androidx.compose.ui.Modifier
//import androidx.compose.ui.draw.shadow
//import androidx.compose.ui.graphics.Color
//import androidx.compose.ui.unit.dp
//import androidx.compose.ui.unit.sp
//
//// Component for displaying the supportive message
//@Composable
//fun SupportMessageBubble(message: String) {
//    var visible by remember { mutableStateOf(false) }
//
//    // Trigger animation when message changes
//    LaunchedEffect(message) {
//        visible = true
//    }
//
//    AnimatedVisibility(
//        visible = visible,
//        enter = fadeIn() + expandVertically()
//    ) {
//        Box(
//            modifier = Modifier
//                .padding(horizontal = 32.dp)
//                .shadow(8.dp, RoundedCornerShape(20.dp))
//                .background(Color.White, RoundedCornerShape(20.dp))
//                .padding(horizontal = 24.dp, vertical = 16.dp)
//        ) {
//            Text(
//                text = message,
//                fontSize = 15.sp,
//                color = Color(0xFF333333),
//                lineHeight = 22.sp
//            )
//        }
//    }
//}

package com.withapp.with.ui.components

import androidx.compose.animation.AnimatedVisibility
import androidx.compose.animation.fadeIn
import androidx.compose.foundation.background
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.shape.GenericShape
import androidx.compose.material3.Text
import androidx.compose.runtime.*
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.shadow
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import com.withapp.with.models.ContextMode
import com.withapp.with.services.GlucoseLevel

enum class SupportLevel { STABLE, BORDERLINE, CONCERNING }

data class SupportMessage(val text: String, val level: SupportLevel) {
    companion object {
        val defaultMessage = SupportMessage("与你同在。", SupportLevel.STABLE)

        fun getMessage(glucoseLevel: GlucoseLevel, context: ContextMode): SupportMessage {
            return when (glucoseLevel) {
                GlucoseLevel.LOW, GlucoseLevel.HIGH -> SupportMessage("请注意你的血糖变化，需要休息一下吗？", SupportLevel.CONCERNING)
                GlucoseLevel.BORDERLINE_LOW, GlucoseLevel.BORDERLINE_HIGH -> SupportMessage("血糖略有波动，记得按时饮食哦。", SupportLevel.BORDERLINE)
                GlucoseLevel.NORMAL -> contextAwareMessage(context)
                GlucoseLevel.UNKNOWN -> defaultMessage
            }
        }

        private fun contextAwareMessage(context: ContextMode): SupportMessage {
            return when (context) {
                ContextMode.RESTAURANT -> SupportMessage("享受美食的同时，别忘了预估碳水哦。", SupportLevel.STABLE)
                ContextMode.WORK -> SupportMessage("工作再忙，也要记得喝水和走动。", SupportLevel.STABLE)
                ContextMode.SOCIAL -> SupportMessage("和朋友聚会很开心吧？我在后台守护你。", SupportLevel.STABLE)
                else -> SupportMessage("你的状态很稳定，继续保持！", SupportLevel.STABLE)
            }
        }
    }
}

@Composable
fun SupportMessageBubble(message: SupportMessage) {
    var isVisible by remember { mutableStateOf(false) }

    LaunchedEffect(message) {
        isVisible = true
    }

    AnimatedVisibility(visible = isVisible, enter = fadeIn()) {
        Box(
            modifier = Modifier
                .padding(horizontal = 24.dp, vertical = 16.dp)
                .shadow(8.dp, GenericShape { size, _ ->
                    val cornerRadius = 40f
                    val arrowSize = 24f
                    val arrowX = size.width / 2f

                    // 纯代码绘制 iOS 风格的带尾巴对话气泡
                    addRoundRect(androidx.compose.ui.geometry.RoundRect(0f, 0f, size.width, size.height - arrowSize, androidx.compose.ui.geometry.CornerRadius(cornerRadius)))
                    moveTo(arrowX - arrowSize, size.height - arrowSize)
                    lineTo(arrowX, size.height)
                    lineTo(arrowX + arrowSize, size.height - arrowSize)
                })
                .background(Color.White)
                .padding(horizontal = 24.dp, vertical = 20.dp),
            contentAlignment = Alignment.Center
        ) {
            Column(horizontalAlignment = Alignment.CenterHorizontally, verticalArrangement = Arrangement.spacedBy(8.dp)) {
                Text(message.text, fontSize = 15.sp, color = Color.Black, textAlign = TextAlign.Center)
                Text("With", fontSize = 11.sp, color = Color(0xFF008080).copy(alpha = 0.7f))
            }
        }
    }
}